bin/p2-rctl/main.go

```diff
@@ -281,32 +290,35 @@ func (r RCtl) RollingUpdate(oldID, newID string, want, need int, deletes bool) {
 	}
 	lock := r.kps.NewUnmanagedLock(session, "")
 
-	u := roll.NewUpdate(roll_fields.Update{
+	signals := make(chan os.Signal, 2)
+	signal.Notify(signals, syscall.SIGTERM, os.Interrupt)
+	quit := make(chan struct{})
+	errs := roll.NewUpdate(roll_fields.Update{
 		OldRC:           rc_fields.ID(oldID),
 		NewRC:           rc_fields.ID(newID),
 		DesiredReplicas: want,
 		MinimumReplicas: need,
 		DeletePods:      deletes,
-	}, r.kps, r.rcs, r.hcheck, r.labeler, r.sched, r.logger, lock)
-
-	if err := u.Prepare(); err != nil {
-		r.logger.WithError(err).Fatalln("Could not prepare update")
-	}
-
-	doneRun := make(chan struct{})
-	go func() {
-		err := u.Run(nil)
-		if err != nil {
-			r.logger.WithError(err).Errorln("Could not run update")
+	}, r.kps, r.rcs, r.hcheck, r.labeler, r.sched, r.logger, lock).Run(quit)
+
+LOOP:
+	for {
+		select {
+		case <-signals:
+			// try to clean up locks on ^C
+			close(quit)
+			r.logger.NoFields().Fatalln("Got signal (manual cleanup of RC locks may be required)")
```

Comment:
Basically the Run() loop is deferring a lock cleanup so that other farms/RUs can claim the lock later. If the program exits too fast the locks might not be released, hence the warning message.

Now that you mention it though, the solution is to not exit right away, but instead to drain the error or session channels to be sure that the child goroutines have exited. Will fix